changelog_unreleased/typescript/14171.md

```diff
@@ -0,0 +1,32 @@
+#### Improve comment print and cursor tracking around type annotation (#14171 by @fisker)
+
+<!-- prettier-ignore -->
+```tsx
+// Input
+let foo/* comment */ : number;
+
+// Prettier stable
+let foo: /* comment */ number;
+
+// Prettier main
+<Same as input>
```

Comment:
I'm probably missing some context, but why keep it same as input? Looks strange. Prettier usually prints a space in front of a comment and doesn't print a space in front of colons.

Reply:
It makes sense to print the comment in front of `:` as it was there input, but the spacing should be like this:
```ts
let foo /* comment */: number;
```

Reply:
Got it.

Reply:
@thorn0 the space after comment is printed by [`printLeadingComment`](https://github.com/prettier/prettier/blob/bf406f772802d702d38bda6a961db0a93529c95b/src/main/comments.js#L456) how would you suggest to remove it? Adjust comment attach? Use `willPrintOwnComments`?

Reply:
Yeah, a tricky problem. It should be printed like a trailing comment, but there is no appropriate node to attach it to as trailing. Need to think about this.

Reply:
Let's keep it and format as 

`let foo /* comment */ : number;`

for now? This is a edge case after all.

Reply:
About the leading space, I also found it's a little bit tricky, the space print inside parent or child also effect the cursor tracking, see [`bb19763`](https://github.com/prettier/prettier/pull/14171/commits/bb197637018eff0e571c81a511e6818e55d15c1c)

I haven't look into it yet, but I think that is a problem should be solved in core.

Reply:
ok